File: python_version/mixed_recall/lowest_common_ancestor/lowest_common_ancestor.py
# lowest common ancestor
# Write a function, lowestCommonAncestor, that takes in the root of
# a binary tree and two values. The function should return the value
# of the lowest common ancestor of the two values in the tree.

# You may assume that the tree values are unique and the tree is non-empty.

# Note that a node may be considered an ancestor of itself.


class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def lowest_common_ancestor(root, val1, val2):

            return 


def get_path(root, val):

    return None


a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')
g = Node('g')
h = Node('h')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g
e.right = h

#      a
#    /    \
#   b      c
#  / \      \
# d   e      f
#    / \
#    g  h
print(lowest_common_ancestor(a, 'd', 'h'))  # b
print(lowest_common_ancestor(a, 'd', 'g'))  # b
print(lowest_common_ancestor(a, 'g', 'c'))  # a
print(lowest_common_ancestor(a, 'b', 'g'))  # b
print(lowest_common_ancestor(a, 'f', 'c'))  # c
# example tree
l = Node('l')
m = Node('m')
n = Node('n')
o = Node('o')
p = Node('p')
q = Node('q')
r = Node('r')
s = Node('s')
t = Node('t')

l.left = m
l.right = n
n.left = o
n.right = p
o.left = q
o.right = r
p.left = s
p.right = t

#        l
#     /     \
#    m       n
#         /    \
#         o     p
#        / \   / \
#       q   r s   t
print(lowest_common_ancestor(l, 'r', 'p'))  # n
print(lowest_common_ancestor(l, 'm', 'o'))  # l
print(lowest_common_ancestor(l, 't', 'q'))  # n
print(lowest_common_ancestor(l, 's', 'p'))  # p

File: java_version/mixed_recall/lowest_common_ancestor/lowest_common_ancestor.java
package mixed_recall.lowest_common_ancestor;
// lowest common ancestor

// Write a function, lowestCommonAncestor, that takes in the root of 
// a binary tree and two values. The function should return the value 
// of the lowest common ancestor of the two values in the tree.

import java.util.*;

// You may assume that the tree values are unique and the tree is non-empty.

// Note that a node may be considered an ancestor of itself.
class Node<T> {
    T val;
    Node<T> left;
    Node<T> right;

    public Node(T val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

public class lowest_common_ancestor {

    public static <T> String lowestCommonAncestor(Node<T> root, String val1, String val2) {

    
        return "";
    }

    public static <T> List<String> getPath(Node<T> root, String val) {

        return new ArrayList<>();
    }

    public static void main(String[] args) {
        Node<String> a = new Node<>("a");
        Node<String> b = new Node<>("b");
        Node<String> c = new Node<>("c");
        Node<String> d = new Node<>("d");
        Node<String> e = new Node<>("e");
        Node<String> f = new Node<>("f");
        Node<String> g = new Node<>("g");
        Node<String> h = new Node<>("h");

        a.left = b;
        a.right = c;
        b.left = d;
        b.right = e;
        c.right = f;
        e.left = g;
        e.right = h;

        // a
        // / \
        // b c
        // / \ \
        // d e f
        // / \
        // g h
        System.out.println(lowestCommonAncestor(a, "d", "h")); // b
        System.out.println(lowestCommonAncestor(a, "d", "g")); // b
        System.out.println(lowestCommonAncestor(a, "g", "c")); // a
        System.out.println(lowestCommonAncestor(a, "b", "g")); // b
        System.out.println(lowestCommonAncestor(a, "f", "c")); // c
        // example tree

        Node<String> l = new Node<>("l");
        Node<String> m = new Node<>("m");
        Node<String> n = new Node<>("n");
        Node<String> o = new Node<>("o");
        Node<String> p = new Node<>("p");
        Node<String> q = new Node<>("q");
        Node<String> r = new Node<>("r");
        Node<String> s = new Node<>("s");
        Node<String> t = new Node<>("t");

        l.left = m;
        l.right = n;
        n.left = o;
        n.right = p;
        o.left = q;
        o.right = r;
        p.left = s;
        p.right = t;

        // l
        // / \
        // m n
        // / \
        // o p
        // / \ / \
        // q r s t
        System.out.println(lowestCommonAncestor(l, "r", "p")); // n
        System.out.println(lowestCommonAncestor(l, "m", "o")); // l
        System.out.println(lowestCommonAncestor(l, "t", "q")); // n
        System.out.println(lowestCommonAncestor(l, "s", "p")); // p

    }
}